### ISSUE TITLE
CA1031: False positive when wrapping exception

### ISSUE THREAD

**Issue by Grauenwolf** [dotnet/roslyn-analyzers#2278]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Package Version

2.9.1

#### Diagnostic ID

CA1031

#### Repro steps

It incorrectly warns about catching System.Exception. As you can see in this example, the caught exception is wrapped and rethrown. 

While not explicitly called out in the documentation, this matches the spirit of the rule which is to not swallow general exceptions.

                    try
                    {
                        value = Convert.ChangeType(value, targetType, CultureInfo.InvariantCulture);
                    }
                    catch (Exception ex)
                    {
                        throw new MappingException($"Cannot map value of type {value.GetType().FullName} to property {property.Name} of type {targetType.Name}.", ex);
                    }

**Comment by mavasani** [comment]
Tagging @m-akinc. Would you be keen to fix this case?

**Comment by m-akinc** [comment]
> Tagging @m-akinc. Would you be keen to fix this case?

This has the potential to get fairly complex. Would we have to handle every way that a thrown object might contain a reference to the original, caught exception? Currently the analyzer reports a diagnostic for all of the following (in addition to the case reported by Grauenwolf):
`catch (Exception ex)
{
    throw ex; // this definitely should not report a diagnostic...
}`
`catch (Exception ex)
{
    var wrapped = new MappingException(..., ex);
    throw wrapped;
}`
`catch (Exception ex)
{
    throw new NotImplementedException();
}`
`catch
{
    throw new NotImplementedException();
}`

If we are okay with *none* of those cases reporting a diagnostic, it's a simple fix: just removing the `operation.Exception == null` part of the condition in `VisitThrow` (https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/DoNotCatchGeneralUnlessRethrown.cs).

I'm not sure I can fix it if we need something more complicated than that.

**Comment by mavasani** [comment]
@m-akinc I think we should be fine with the simple fix for now. We probably want to ensure that the end point of the catch clause is not reachable by invoking [AnalyzeControlFlow](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/SemanticModel.cs,662) on the catch clause and verifying that [EndPointIsReachable](http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/ControlFlowAnalysis.cs,29) is `false`.

**Comment by m-akinc** [comment]
@mavasani I ran into problems using AnalyzeControlFlow. Specifically, for VB, AnalyzeControlFlow does not like being passed a `CatchBlockSyntax`. It throws an exception that 'statement is not an StatementSyntax'.

Also, it seems like that approach has its own issues -- the catch block has a `break` or `return`, wouldn't that potentially give a false negative?

I experimented with `AnalyzeDataFlow`, hoping that it would indicate whether the caught exception instance was being emitted, but that seemed not to work. Maybe throwing is not covered by the `DataFlowsOut` analysis.

Anyway, I've created a PR #2354 for the simplest fix.
